### ISSUE TITLE
Getting a slice out of an ArrayView: cannot return value referencing function parameter

### ISSUE THREAD

**Issue by SuperFluffy** [rust-ndarray/ndarray#621]
I am trying to write a function that interprets an `ArrayView` as row major order, but Rust doesn't like me extracting the slice underlying the `ArrayView`. Here is a mimimal example showing the problem:

```rust
fn as_standard_view<'a, T>(a: ArrayView<'a, T, Ix2>) -> ArrayView<'a, T, Ix1> {
    let slice = a.as_slice_memory_order().unwrap();
    aview1(slice)
}
```

This is the associated error: 

```
error[E0515]: cannot return value referencing function parameter `a`
  --> simulate/src/system_serialization.rs:58:5
   |
57 |     let slice = a.as_slice_memory_order().unwrap();
   |                 - `a` is borrowed here
58 |     aview1(slice)
   |     ^^^^^^^^^^^^^ returns a value referencing data owned by the current function
```

I do understand the error, but I am not sure how to work around it. After all, the `slice` is held by the `ArrayView`, but not the data itself.

Any idea how to construct thay `ArrayView` without rust thinking that I am trying to reference something inside the function?

**Comment by SuperFluffy** [comment]
Not 100% sure, but I believe this comes from `as_slice` and `as_slice_memory_order` tying the lifetime of the resulting slice not to the lifetime parameter `'a` of the `ArrayView`, but to the lifetime parameter of the borrowed `ArrayView`, see the implementation here:

```rust
    pub fn as_slice_memory_order(&self) -> Option<&[A]>
    where
        S: Data,
    {
        if self.is_contiguous() {
            unsafe {
                Some(slice::from_raw_parts(self.ptr, self.len()))
            }
        } else {
            None
        }
    }
```

**Comment by SuperFluffy** [comment]
I think one approach is to just `unsafe`ly create a slice, mimicking what `as_slice_memory_order` does, but binding the lifetime to the parameter inside `ArrayView`:

```rust
fn as_standard_view<'a, 'b: 'a, T>(a: &'a ArrayView<'b, T, Ix2>) -> ArrayView<'b, T, Ix1> {
    let a_slice = a.as_slice_memory_order().unwrap();
    let a_ptr = a_slice.as_ptr();
    let new_slice = unsafe { slice::from_raw_parts(a_ptr, a_slice.len()) };
    aview1(new_slice)
}
```

I think the proper fix it to have specific implementations of `as_slice{,_memory_order` depending on whether we have an `Array` or an `ArrayView`.

**Comment by jturner314** [comment]
It sounds like you need an `.into_slice_memory_order()` method for `ArrayView` analogous to the existing [`.into_slice()`](https://docs.rs/ndarray/0.12.1/ndarray/type.ArrayView.html#method.into_slice) method. Please feel free to submit a PR adding this for `ArrayView` and `ArrayViewMut`. (It should go in `impl_views.rs` below the existing `into_slice` methods.)

**Comment by SuperFluffy** [comment]
Thanks for the headsup on `.into_slice()`. Bit of a discoverability issue here. :)

What's your opinion on providing a specific implementation of `as_slice` and `as_slice_memory_order` to fix the lifetime bounds?

**Comment by jturner314** [comment]
> Thanks for the headsup on `.into_slice()`. Bit of a discoverability issue here. :)

Yeah, it would be good to add a note to the `.as_slice()` and `.as_slice_mut()` docs referring to that method. (A PR would be welcome. :) You may also be interested in [this section of the docs](https://docs.rs/ndarray/0.12.1/ndarray/struct.ArrayBase.html#conversions-between-arrays-and-vecsslicesscalars) that was added relatively recently.

> What's your opinion on providing a specific implementation of `as_slice` and `as_slice_memory_order` to fix the lifetime bounds?

I don't think we should do this because then `as_slice` would no longer be implemented for generic `S: Data`; it would have to be implemented for each specific `S`. (This would be repetitive and would be inconvenient for client code that uses generic `S: Data`.) `.into_slice()` and `.into_slice_memory_order()` can provide the necessary functionality for `ArrayView` and `ArrayViewMut`.
